src/preconditioner.jl

```diff
@@ -1,13 +1,42 @@
-import Base: \, *, A_ldiv_B!, A_mul_B!
-
-struct Preconditioner
-    ml::MultiLevel
+struct Preconditioner{ML<:MultiLevel}
+    ml::ML
+    init::Symbol
 end
+Preconditioner(ml) = Preconditioner(ml, :zero)
 
 aspreconditioner(ml::MultiLevel) = Preconditioner(ml)
 
-\(p::Preconditioner, b) = solve(p.ml, b, maxiter = 1, tol = 1e-12)
-*(p::Preconditioner, b) = p.ml.levels[1].A * x
+@static if VERSION < v"0.7-"
+    import Base: \, *, A_ldiv_B!, A_mul_B!
+    function A_ldiv_B!(x, p::Preconditioner, b)
+        if p.init == :zero
+            x .= 0
+        else
+            x .= b
+        end
+        solve!(x, p.ml, b, maxiter = 1, calculate_residual = false)
```

Comment:
maintain `tol = 1e-12`

Reply:
Because `calculate_residual` is false, `tol` will be ignored and since `tol` is way too small, it will never really lead to the termination of the loop. Computing the tolerance is expensive so blindly running one iteration of the AMG solver is more efficient. Again this is what is going to happen anyways unless the residual norm is less than `tol`.